shell/platform/linux/fl_view.cc

```diff
@@ -152,6 +153,61 @@ static gboolean run_flutter_engine(FlView* self) {
   return TRUE;
 }
 
+static gboolean fl_view_send_pointer_button_event(FlView* self,
+                                                  GdkEventButton* event) {
+  FlutterPointerEvent fl_event = {};
+  fl_event.struct_size = sizeof(fl_event);
+  fl_event.timestamp = event->time * 1000;
+
+  int64_t button;
+  switch (event->button) {
+    case 1:
+      button = kFlutterPointerButtonMousePrimary;
+      break;
+    case 2:
+      button = kFlutterPointerButtonMouseMiddle;
+      break;
+    case 3:
+      button = kFlutterPointerButtonMouseSecondary;
+      break;
+    default:
+      return FALSE;
+  }
+  int old_button_state = self->button_state;
+  if (event->type == GDK_BUTTON_PRESS) {
+    // Drop the event if Flutter already thinks the button is down
+    if ((self->button_state & button) != 0)
+      return FALSE;
+    self->button_state ^= button;
+
+    if (old_button_state == 0)
+      fl_event.phase = kDown;
+    else
+      fl_event.phase = kMove;
+  } else if (event->type == GDK_BUTTON_RELEASE) {
+    // Drop the event if Flutter already thinks the button is up
+    if ((self->button_state & button) == 0)
+      return FALSE;
+    self->button_state ^= button;
+
+    if (self->button_state == 0)
+      fl_event.phase = kUp;
+    else
+      fl_event.phase = kMove;
+  }
+
+  if (self->flutter_engine == nullptr)
```

Comment:
It's done lower down because we want to keep the state correct, even if we're not sending the event to the engine - in case the engine is changed due to a later code change.